ip-utils.c

```diff
@@ -232,23 +200,84 @@ const char *janus_network_address_string_from_buffer(const janus_network_address
-	if(iface == NULL)
-		return -1;
 	janus_network_query_config q;
 	/* Let's see if iface is an IPv4 address, an IPv6 address, or possibly an interface name */
-	if(janus_network_prepare_device_query(iface,
-			janus_network_query_options_ipv4 | janus_network_query_options_ipv6 | janus_network_query_options_name, &q)) {
+	int res = janus_network_prepare_device_query(iface,
+			janus_network_query_options_ipv4 | janus_network_query_options_ipv6 | janus_network_query_options_name, &q);
+	if(res != 0) {
 		/* None of them..? */
-		return -1;
+		return res;
 	}
 	const struct ifaddrs *found = janus_network_query_devices(ifas, &q);
 	if(!found || janus_network_get_device_address(found, result)) {
 		/* Couldn't find anything on iface */
-		return -1;
+		return -EINVAL;
 	}
 	/* Done */
 	return 0;
 }
+
+int janus_network_detect_local_ip(janus_network_query_options addr_type, janus_network_address *result) {
+	if(result == NULL)
+		return -EINVAL;
+	janus_network_address_nullify(result);
+	gboolean found = FALSE;
+	int fd = -1;
+	if(addr_type == janus_network_query_options_ipv4 || addr_type == janus_network_query_options_any_ip) {
+		/* Let's try IPv4 (FIXME Should probably use other internal methods) */
+		struct sockaddr_in addr;
+		socklen_t len;
+		fd = socket(AF_INET, SOCK_DGRAM, 0);
+		if(fd > -1) {
+			addr.sin_family = AF_INET;
+			addr.sin_port = htons(1);
+			inet_pton(AF_INET, "1.2.3.4", &addr.sin_addr);
+			if(connect(fd, (const struct sockaddr*)&addr, sizeof(addr)) > -1) {
+				len = sizeof(addr);
+				if(getsockname(fd, (struct sockaddr*)&addr, &len) > -1) {
+					result->family = AF_INET;
+					if(janus_ip_copy_ipv4((struct sockaddr_in *)&addr, &result->ipv4) == 0) {
+						found = TRUE;
+					}
+				}
+			}
+		}
+	}
+	if(fd != -1)
+		close(fd);
+	if(!found && (addr_type == janus_network_query_options_ipv6 || addr_type == janus_network_query_options_any_ip)) {
+		/* Let's try IPv6 (FIXME Should probably use other internal methods) */
+		struct sockaddr_in6 addr;
+		socklen_t len;
+		int fd = socket(AF_INET6, SOCK_DGRAM, 0);
+		if(fd > -1) {
+			addr.sin6_family = AF_INET6;
+			addr.sin6_port = htons(1);
+			inet_pton(AF_INET6, "::1.2.3.4", &addr.sin6_addr);
+			if(connect(fd, (const struct sockaddr*)&addr, sizeof(addr)) > -1) {
+				len = sizeof(addr);
+				if(getsockname(fd, (struct sockaddr*)&addr, &len) > -1) {
+					result->family = AF_INET6;
+					if(janus_ip_copy_ipv6((struct sockaddr_in6 *)&addr, &result->ipv6) == 0) {
+						found = TRUE;
+					}
+				}
+			}
+		}
+	}
+	if(fd != -1)
+		close(fd);
+	return found ? 0 : -EINVAL;
+}
+
+char *janus_network_detect_local_ip_as_string(janus_network_query_options addr_type) {
+	janus_network_address addr;
+	janus_network_address_string_buffer buf;
+    int res = janus_network_detect_local_ip(addr_type, &addr) || janus_network_address_to_string_buffer(&addr, &buf);
+	if(res != 0)
+		return NULL;
+	return g_strdup(janus_network_address_string_from_buffer(&buf));
```

Comment:
It was ternary in the code you proposed: I preferred to split it to make it better readable.